src/data/pfun.lean

```diff
@@ -448,4 +448,45 @@ ext $ λ _ _, by simp only [comp_apply, part.bind_comp]
 lemma coe_comp (g : β → γ) (f : α → β) : ((g ∘ f : α → γ) : α →. γ) = (g : β →. γ).comp f :=
 ext $ λ _ _, by simp only [coe_val, comp_apply, part.bind_some]
 
+/-- Product of partial functions. -/
+def prod_lift (f : α →. β) (g : α →. γ) : α →. β × γ :=
+λ x, ⟨(f x).dom ∧ (g x).dom, λ h, ((f x).get h.1, (g x).get h.2)⟩
+
+@[simp] lemma dom_prod_lift (f : α →. β) (g : α →. γ) :
+  (f.prod_lift g).dom = {x | (f x).dom ∧ (g x).dom} := rfl
+
+lemma get_prod_lift (f : α →. β) (g : α →. γ) (x : α) (h) :
+  (f.prod_lift g x).get h = ((f x).get h.1, (g x).get h.2) := rfl
+
+@[simp] lemma prod_lift_apply (f : α →. β) (g : α →. γ) (x : α) :
+  f.prod_lift g x = ⟨(f x).dom ∧ (g x).dom, λ h, ((f x).get h.1, (g x).get h.2)⟩ := rfl
+
+lemma mem_prod_lift {f : α →. β} {g : α →. γ} {x : α} {y : β × γ} :
+  y ∈ f.prod_lift g x ↔ y.1 ∈ f x ∧ y.2 ∈ g x :=
+by { simp [prod_lift], tidy }
+
+/-- Product of partial functions. -/
+def prod_map (f : α →. γ) (g : β →. δ) : α × β →. γ × δ :=
+λ x, ⟨(f x.1).dom ∧ (g x.2).dom, λ h, ((f x.1).get h.1, (g x.2).get h.2)⟩
+
+@[simp] lemma dom_prod_map (f : α →. γ) (g : β →. δ) :
+  (f.prod_map g).dom = {x | (f x.1).dom ∧ (g x.2).dom} := rfl
+
+lemma get_prod_map (f : α →. γ) (g : β →. δ) (x : α × β) (h) :
+  (f.prod_map g x).get h = ((f x.1).get h.1, (g x.2).get h.2) := rfl
+
+@[simp] lemma prod_map_apply (f : α →. γ) (g : β →. δ) (x : α × β) :
+  f.prod_map g x = ⟨(f x.1).dom ∧ (g x.2).dom, λ h, ((f x.1).get h.1, (g x.2).get h.2)⟩ := rfl
+
+lemma mem_prod_map {f : α →. γ} {g : β →. δ} {x : α × β} {y : γ × δ} :
+  y ∈ f.prod_map g x ↔ y.1 ∈ f x.1 ∧ y.2 ∈ g x.2 :=
+by { simp [prod_map], tidy }
+
+@[simp] lemma prod_map_id_id : (pfun.id α).prod_map (pfun.id β) = pfun.id _ :=
```

Comment:
You might like to look at the lemmas we have for products in a category and see how many (if any) you'd like to adapt - this and the below are there for example

Reply:
Yeah I got those two from the category ones, actually. I am not sure how much I am supposed to duplicate.

Reply:
My guess is just add the ones you need, copying the names from the category ones but leaving out any which you don't find useful for your application

Reply:
Then I think the existing two are all I need, at least for now.